### ISSUE TITLE
Task queue

### ISSUE THREAD

**Issue by cahirwpz** [cahirwpz/wifire-os#36]
**Depends on #32**

A task queue is needed to run bottom halves of interrupt handlers. A top half of interrupt is executed during interrupt request handling. Time consuming and possibly blocking work is deferred to be executed within high priority kernel thread. Interface proposal:

```
typedef void task_fn_t(void *data);

struct task_queue {
    STAILQ_ENTRY(task_queue) next;
    work_fn_t *func;
    void *data;
};

void task_queue_add(struct task_queue *queue, task_fn_t *task, void *data);
void task_queue_run(struct task_queue *queue);
```

Find similar functionality in other kernels (Minix / *BSD / Linux) and discuss them here. Propose your data structures and functions.

**Comment by Killavus** [comment]
So what we basically want to do is to accept every intr handler asap, but if it needs more work we'd like to keep such work within a queue and process it later (if kernel is busy)?

I'd like to take a look at how BSD does it. I believe "your data structures" means "steal it if it is possible and feasible for us"? I'd rather want to minimize the 'talk part' and focus on delivering it since I failed with list implementations hard.

**Comment by cahirwpz** [comment]
Nope. Please read the justification behind splitting interrupt handler into two parts. I hoped it had been well described during a seminar talk. There're several good articles about that. 

Research and discussion is very important - don't neglect it. I have no time to investigate every implementation... you should convince me that stealing / implementing one particular solution is a good idea.

**Comment by Killavus** [comment]
OK. I'll refresh the talk from seminar (because I just remembered this particular part) and will research on the topic. Thanks!

**Comment by cahirwpz** [comment]
Any progress?

**Comment by cahirwpz** [comment]
The scope of this task is unclear to me. It has to be redefined from scratch.
